### PR TITLE
[SPIR-V] Add cl_khr_kernel_clock / SPV_KHR_shader_clock extension

### PR DESCRIPTION
**Pull request by svenvh**
Recognize `cl_khr_kernel_clock` builtins and translate them to `OpReadClockKHR` instructions.  The `Scope` operand is deduced from the builtin function name.

spirv-val does not pass yet due to OpReadClockKHR only supporting the valid scopes for Vulkan (Device and Subgroup, but not Workgroup), so leave validation disabled with a TODO.

Provisional extension description: https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#cl_khr_kernel_clock .
Builtins: https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_C.html#kernel-clock-functions .
SPIR-V environment: https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Env.html#_cl_khr_kernel_clock .

### CODE REVIEW

File: llvm/test/CodeGen/SPIRV/extensions/SPV_KHR_shader_clock/shader_clock.ll [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+; CHECK: OpReadClockKHR [[ulong]] [[uint_1]]
+; CHECK: OpReadClockKHR [[ulong]] [[uint_2]]
+; CHECK: OpReadClockKHR [[ulong]] [[uint_3]]
+; CHECK: OpReadClockKHR [[v2uint]] [[uint_1]]
+; CHECK: OpReadClockKHR [[v2uint]] [[uint_2]]
+; CHECK: OpReadClockKHR [[v2uint]] [[uint_3]]
+
+; ModuleID = '<stdin>'
+target datalayout = "e-p:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-G1"
+target triple = "spir-unknown-unknown"
```
Comment (VyacheslavLevytskyy) [inline llvm/test/CodeGen/SPIRV/extensions/SPV_KHR_shader_clock/shader_clock.ll:24]: Let's remove those 3 lines please, I think there is no need to keep them in the test

File: llvm/test/CodeGen/SPIRV/extensions/SPV_KHR_shader_clock/shader_clock.ll [hunk 2]
```diff
@@ -0,0 +1,87 @@
+; RUN: not llc -O0 -mtriple=spirv32-unknown-unknown %s -o %t.spvt 2>&1 | FileCheck %s --check-prefix=CHECK-ERROR
+; RUN: llc -O0 -mtriple=spirv32-unknown-unknown --spirv-ext=+SPV_KHR_shader_clock %s -o - | FileCheck %s
+; TODO: %if spirv-tools %{ llc -O0 -mtriple=spirv64-unknown-unknown --spirv-ext=+SPV_KHR_shader_clock %s -o - -filetype=obj | spirv-val %}
```
Comment (VyacheslavLevytskyy) [inline llvm/test/CodeGen/SPIRV/extensions/SPV_KHR_shader_clock/shader_clock.ll:3]: llc and spirv-val runs differ in spirv32 and spirv64

**Comment by VyacheslavLevytskyy** [comment]
@svenvh Thank you and very glad to see this contribution! My only suggestions are about the test case, with a general goal to make it slightly cleaner, devoid of unimportant details -- otherwise PR looks good.

File: llvm/test/CodeGen/SPIRV/extensions/SPV_KHR_shader_clock/shader_clock.ll [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+; CHECK: OpReadClockKHR [[v2uint]] [[uint_2]]
+; CHECK: OpReadClockKHR [[v2uint]] [[uint_3]]
+
+; ModuleID = '<stdin>'
+target datalayout = "e-p:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-G1"
+target triple = "spir-unknown-unknown"
+
+; Function Attrs: convergent norecurse nounwind
+define dso_local spir_kernel void @test_clocks(ptr addrspace(1) nocapture noundef writeonly align 8 %out64, ptr addrspace(1) nocapture noundef writeonly align 8 %outv2) local_unnamed_addr #0 !kernel_arg_addr_space !3 !kernel_arg_access_qual !4 !kernel_arg_type !5 !kernel_arg_base_type !6 !kernel_arg_type_qual !7 {
+entry:
```
[the remark this answers is not in the document]
Comment (svenvh) [reply llvm/test/CodeGen/SPIRV/extensions/SPV_KHR_shader_clock/shader_clock.ll:23]: Done.

**Review by VyacheslavLevytskyy — APPROVED** [review]

**Pull request merged**